store/tikv/2pc.go

```diff
@@ -214,14 +228,16 @@ func (c *twoPhaseCommitter) doActionOnKeys(bo *Backoffer, action twoPhaseCommitA
 		sizeFunc = c.keyValueSize
 	}
 	// Make sure the group that contains primary key goes first.
-	batches = appendBatchBySize(batches, firstRegion, groups[firstRegion], sizeFunc, txnCommitBatchSize)
-	delete(groups, firstRegion)
+	if len(groups[primaryRegion]) > 0 {
+		batches = appendBatchBySize(batches, primaryRegion, groups[primaryRegion], sizeFunc, txnCommitBatchSize)
+		markBatchAsPrimary(batches, c.primary())
+		delete(groups, primaryRegion)
+	}
 	for id, g := range groups {
 		batches = appendBatchBySize(batches, id, g, sizeFunc, txnCommitBatchSize)
 	}
 
-	firstIsPrimary := bytes.Equal(keys[0], c.primary())
-	if firstIsPrimary && (action == actionCommit || action == actionCleanup) {
+	if batches[0].primary && (action == actionCommit || action == actionCleanup) {
```

Comment:
The primary batch may still not in the first place.

Reply:
Forgot it...